Tracing.gprc interceptor

andreimatei opened the pull request.

andreimatei committed:
util/tracing: move Recording to tracingpb

The Recording type was defined identically in both the tracing and the
tracingpb package. This was confusing. This patch removes the tracing
definition, leaving only tracingpb.Recording.

Release note: None

andreimatei committed:
util/tracing: move the gRPC interceptors to a new package

The gRPC interceptors for tracing were defined in the tracing package.
This patch moves them to a dedicated packages, so that they can depend
on roachpb. The tracing package is at the base of everything, so we
should keep it small.

Release note: None

andreimatei committed:
rpc: run grpc interceptors for local RPCs

Before this patch, the gRPC interceptors that we normally run on the
client and the server were not being run for "RPCs" to the "Internal"
sevice going to the local node. Such RPCs are done through the
localClientAdapter, which bypasses gRPC for performance reasons. Not
running these interceptors is bad - they do useful things around
creating tasks, tracing, authentication/authorization.
Some of the things done by the interceptors was done separately and
redundantly in random places to account for the RPC local case.

This patch unifies the paths of local and remote RPCs by having the
localClientAdapter run the client and server interceptors. Care was
taken for the performance regression on the local path; still, one
new allocation is imposed by the client interceptor interface for
(unary) RPC calls: the result of the RPC needs to be allocated before
the RPC is made, and the proto that the server returns needs to be
copied over.
I've run some benchmarks with this change and it didn't show anything.

Node.batchInternal was simplified slightly to not create a task, since
the server interceptor now does it in all cases. Before, creating the
task was redundant in the non-local RPC case.

Fixes #74732

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/76263)
<!-- Reviewable:end -->

andreimatei committed:
rpc,kvclient: cull the internalClientAdapter

rpc.internalClientAdapter is used to call into the roachpb.Internal gRPC
service of the local Node without actually going through gRPC. This
structure was implementing all of the ever-growing Internal interface,
even though it was only ever used for exactly two methods - the ones
used by the DistSender (Batch and RangeFeed). This patch extracts the
sub-interface of Internal needed by the DistSender, and trims the
internalClientAdapter to only implement the respective two methods.

There are multiple benefits in doing this trimming:
1) The internalClientAdapter had too much dead code. In particular for
streaming methods, the implementation is non-trivial. The respective
code is mostly copy-paste, but still.
2) The code in the internalClientAdapter, as written, was
insufficient(*) as it was not setting a magic auth key on the ctx
required in order to pass the Authentication gRPC server interceptor. In
the case of the two methods used by the DistSender, our
nodedialer.Dialer does an awkward thing where it create a ctx with the
magic key and returns that ctx over several levels to the grpcTransport,
which use it for the upcoming gRPC call. This patch removes this awkward
protocol by having the internalClientAdaptor populate the auth key in
the ctx.

(*) Full disclosure - the respective magic key is not actually needed
since we don't currently run the gRPC interceptors for these local RPCs.
But I'm trying to change that.

Release note: None
